src/EFCore.PG/Query/Internal/NpgsqlEvaluatableExpressionFilter.cs

```diff
@@ -0,0 +1,28 @@
+﻿using System.Linq.Expressions;
+using System.Reflection;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Metadata;
+using NpgsqlTypes;
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal
+{
+    public class NpgsqlEvaluatableExpressionFilter : RelationalEvaluatableExpressionFilter
```

Comment:
> Is this because trying to call the full-text search methods on a local vector+query will throw a NotSupportedException

Yes.

It's because of the query compiler. It tries to eliminate all calls in LINQ expressions and convert them to constant expressions (evaluated client side) when they don't reference any column, but in case of full text search functions, there is simply no implementation to run client side, so the behavior is incorrect as the expressions would throw and I must force it to evaluate them server-side. 

```RelationalEvaluatableExpressionFilter``` comes from ```EvaluatableExpressionFilter``` which is a public API from Relinq (not EF Core), so this API will never go away, if EF ever make ```RelationalEvaluatableExpressionFilter``` internal or private, we can easily update and wrap/proxy the original object without any redesign (unless they rebase the entire EF core framework on something other than Relinq but that is highly unlikely). So I think my use of it is the 100% intended use case by Relinq folks and it should be safe to keep.

Sql Server folks have made a choice to let exceptions be thrown if the expression doesn't reference any column (like Freetext) because they guess nobody will use it that way (as 'compute'), but in my opinion that is a bug, or perhaps there is no way in Sql Server to call such functions without referencing columns. Their handling of DateTime.Now is the bare minimum implementation and I think more methods should be added to the filter (anything that doesn't have a client-side implementation).

Allowing users to use the relational store for 'compute' may not be an orthodox use case but there might be some fringe uses for it. I don't see why we should arbitrarily create an incorrect behavior/implementation hole that was not there in EF6 (which can handle such queries just fine) just because the Sql server people decided to do that (or maybe didn't test that case).

Let me know what you think :-).

Reply:
See https://github.com/aspnet/EntityFrameworkCore/issues/2667 @roji when they started using this API from relinq.

Reply:
@ajcvickers, @divega could we get a bit of guidance here?

To summarize the issue, @rwasef1830 is finalizing the full-text support, and proposes to add NpgsqlEvaluatableExpressionFilter in order to ensure that the relevant functions always execute on the server. This is because these functions, like SQL Server's `FREETEXT`, have no client implementation and would always throw. The main (only?) use case would be to allow running these database functions on client-local data.

My personal opinion is that the usage scenario isn't important enough, but please see @rwasef1830's comment above (and mine). It would be interesting to hear your thoughts on this.

Reply:
@roji we discussed this with @ajcvickers and @smitpatel (@anpete was not part of the discussion and he may also have something relevant to add). 

In summary, we believe that the behavior @rwasef1830 is proposing is reasonable. We currently are not doing this with the SQL Server provider but in the future we could stop throwing the exception and start issuing new queries.

It will result in N+1 queries being sent to the server, but as long as there are appropriate client evaluation warnings are being reported this is not different from other cases in which other constructs (e.g. subqueries) result into N+1.

There might be a problem with async queries: ideally each individual function evaluation should be evaluated asynchronously not block. I actually haven't looked at the proposed implementation so I am not sure if this is already handled. If it isn't, this is something that I believe @anpete knows how to handle the async composition and he may be able to provide some pointers.

I have created https://github.com/aspnet/EntityFrameworkCore/issues/11466 to track the idea of doing this more often.

Reply:
OK, thanks for the guidance. I'll merge this PR as-is, and we'll change the implementation when #11466 is implemented (possibly with a better way to add methods as @anpete suggests).